### PR TITLE
Adding NST with AdaIN

### PR DESCRIPTION
**Pull request by ariG23498**
In this PR we try to implement Neural Style Transfer with [Adaptive Instance Normalization](https://arxiv.org/abs/1703.06868).

[Colab Notebook](https://colab.research.google.com/drive/1mn9WrC6eT3hCVlVBGoILQScsP5krQibV?usp=sharing)
[Hugging Face demo](https://huggingface.co/spaces/ariG23498/nst)

### CODE REVIEW

File: examples/vision/adain.py [hunk 1]
```diff
@@ -0,0 +1,675 @@
...
+
+
+class NST(tf.keras.Model):
+    def __init__(self, encoder, decoder, ada_in, loss_net, lamb):
+        super().__init__()
+        self.encoder = encoder
+        self.decoder = decoder
+        self.ada_in = ada_in
+        self.loss_net = loss_net
+        self.lamb = lamb
```
Comment (LukeWood) [inline examples/vision/adain.py:466]: if this is lambda, let's just call it `lambda`.  The lamb abbreviation can cause confusion.

File: examples/vision/adain.py [hunk 2]
```diff
@@ -0,0 +1,675 @@
...
+    def train_step(self, inputs):
+        style, content = inputs
+
+        # Initialize the content and style loss.
+        loss_content = 0.0
+        loss_style = 0.0
+
+        with tf.GradientTape() as tape:
+            # Encode the style and content image.
+            style_enc = self.encoder(style)
```
Comment (LukeWood) [inline examples/vision/adain.py:485]: enc -> encoded

File: examples/vision/adain.py [hunk 3]
```diff
@@ -0,0 +1,675 @@
...
+            content_enc = self.encoder(content)
+
+            # Compute the AdaIN feature map.
+            t = self.ada_in(style=style_enc, content=content_enc)
```
Comment (LukeWood) [inline examples/vision/adain.py:489]: same here, enc -> encoded.

Comment (LukeWood) [inline examples/vision/adain.py:489]: anything more descriptive we can call `t`?  it's ok if not.

File: examples/vision/adain.py [hunk 4]
```diff
@@ -0,0 +1,675 @@
...
+
+            # Generate the neural style transferred image.
+            recons_image = self.decoder(t)
+
+            # Compute the losses.
+            recons_vgg_feats = self.loss_net(recons_image)
```
Comment (LukeWood) [inline examples/vision/adain.py:495]: recons->reconstructed

File: examples/vision/adain.py [hunk 5]
```diff
@@ -0,0 +1,675 @@
...
+        with tf.GradientTape() as tape:
+            # Encode the style and content image.
+            style_enc = self.encoder(style)
+            content_enc = self.encoder(content)
+
+            # Compute the AdaIN feature map.
+            t = self.ada_in(style=style_enc, content=content_enc)
+
+            # Generate the neural style transferred image.
+            recons_image = self.decoder(t)
```
Comment (LukeWood) [inline examples/vision/adain.py:492]: recons->reconstructed all over

File: examples/vision/adain.py [hunk 6]
```diff
@@ -0,0 +1,675 @@
...
+
+            # Compute the losses.
+            recons_vgg_feats = self.loss_net(recons_image)
+            style_vgg_feats = self.loss_net(style)
```
Comment (LukeWood) [inline examples/vision/adain.py:496]: feats -> features

File: examples/vision/adain.py [hunk 7]
```diff
@@ -0,0 +1,675 @@
...
+            loss_content = self.loss_fn(t, recons_vgg_feats[-1])
+            for inp, out in zip(style_vgg_feats, recons_vgg_feats):
```
Comment (LukeWood) [inline examples/vision/adain.py:498]: feats -> features

File: examples/vision/adain.py [hunk 8]
```diff
@@ -0,0 +1,675 @@
...
+                mean_inp, std_inp = get_mean_std(inp)
```
Comment (LukeWood) [inline examples/vision/adain.py:499]: maybe inp => features?

File: examples/vision/adain.py [hunk 9]
```diff
@@ -0,0 +1,675 @@
...
+            "style_loss": self.style_loss_tracker.result(),
+            "content_loss": self.content_loss_tracker.result(),
+            "total_loss": self.total_loss_tracker.result(),
+        }
+
+    def test_step(self, inputs):
+        style, content = inputs
+        loss_content = 0.0
+        loss_style = 0.0
+        style_enc = self.encoder(style)
```
Comment (LukeWood) [inline examples/vision/adain.py:526]: same comments here as train_step

File: examples/vision/adain.py [hunk 10]
```diff
@@ -0,0 +1,675 @@
...
+        ]
+
+
+"""
+## Train Monitor callback
+
+This callback is used to visualize the style transfer capability of
+the model on each epoch. The objective of style transfer is not
+quantified properly and is mostly upto the audience to call is good
+or bad. For this reason, visualization falls into a key aspect of
```
Comment (LukeWood) [inline examples/vision/adain.py:565]: -> change to:

visualization is a key aspect of training the model

File: examples/vision/adain.py [hunk 11]
```diff
@@ -0,0 +1,675 @@
...
+training the model.
+"""
+
+test_style, test_content = next(iter(test_ds))
+
+
+class TrainMonitor(tf.keras.callbacks.Callback):
+    def on_epoch_end(self, epoch, logs=None):
+        # Encode the style and content image.
+        test_style_enc = self.model.encoder(test_style)
```
Comment (LukeWood) [inline examples/vision/adain.py:575]: same renames, encoded

File: examples/vision/adain.py [hunk 12]
```diff
@@ -0,0 +1,675 @@
...
+        ax_s.set_title("Style Image")
+        ax_c.imshow(c_im)
+        ax_c.set_title("Content Image")
+        ax_r.imshow(r_im)
+        ax_r.set_title("NST Image")
+
+"""
+# Conclusion
+
+Adaptive Instance Normalization is an important step in the direction
```
Comment (LukeWood) [inline examples/vision/adain.py:662]: we can remove "is an important step in the direction of Neural Style Transfer research because it" resulting in...

`Adaptive Instance Normalization allows arbitrary style
transfer in real time`

File: examples/vision/adain.py [hunk 13]
```diff
@@ -0,0 +1,675 @@
...
+of Neural Style Transfer research beacuse it allows arbitrary style
+transfer in real time. It is also important to note that the novel
+proposition of the authors is to achive this only by aligning the
```
Comment (LukeWood) [inline examples/vision/adain.py:665]: achive -> achieve

**Review by LukeWood — CHANGES_REQUESTED** [review]
Thank you for your contribution.  This is a really great example, I left some comments.  Please address them and reply back.

I really love real time Neural Style Transfer (I published https://imstyle.app awhile back...) so it's great to see some extra examples using it.

**Comment by LukeWood** [comment]
Also note you will need to format your file after the changes to get the CI workflow to pass.

```
black examples/vision/adain.py
```
should work

File: examples/vision/adain.py (hunk 1 above)
Comment (ritwikraha) [reply examples/vision/adain.py:466]: We think ```style_weight``` is more accurate here, so we replaced ```lamb``` with that.

File: examples/vision/adain.py (hunk 3 above)
Comment (ritwikraha) [reply examples/vision/adain.py:489]: The term ```t``` (target feature maps) is taken from the paper itself
image
So we think its best to leave it as ```t```

**Review by ritwikraha — COMMENTED** [review]
@LukeWood thanks for the detailed review. We have fixed most of the things you mentioned and will push the changes for review shortly

File: examples/vision/adain.py [hunk 14]
```diff
@@ -0,0 +1,675 @@
+"""
```
[the remark this answers is not in the document]
Comment (ariG23498) [reply examples/vision/adain.py:1]: Thank you for your kind words!

**Comment by ariG23498** [comment]
@LukeWood Thanks for your thorough review. We have taken care of each of the suggestions. After the changes we have used the black formatting yet the CI/CD pipeline does not seem to work for us. Could you help here?

**Comment by LukeWood** [comment]
Looks good!  Please run:
```
cd scripts
python3 autogen.py add_example vision/adain.py
```
to add the generated files

**Comment by ariG23498** [comment]
@LukeWood we have added the generated files.

**Review by LukeWood — APPROVED** [review]

**Comment by LukeWood** [comment]
@ariG23498 are the images in examples/vision/img/adain/adain_25_0.png correct?  I see a deeply red image for every style/every content image.  Did something go wrong in the figure generation?

**Review by LukeWood — CHANGES_REQUESTED** [review]
Let's make sure those figures are correct.  If needed you can fit the model for longer locally and host the resulting diagrams on IMGUR.

**Comment by LukeWood** [comment]
Ah -- it's because we only ran for 1 epoch.  I wonder.... do you think it's more informative/impressive to link to results from 30 epochs?

**Comment by ariG23498** [comment]
Hey @LukeWood thank you for observing this.

In the introduction, we have added the output for a 30 epoch trained model. The image is 

img

We have also added a hugging face demo for style transfer with the pretrained models. The link is: https://huggingface.co/spaces/ariG23498/nst

I think this should be enough for the reader to familiarise with the output of a trained mode. What do you think?

**Comment by LukeWood** [comment]
Ah fantastic - thank you.

Would you mind running `python3 autogen.py add_example vision/adain.py`
 one more time?  I updated one more link :)

Then I think we are good!

**Comment by ariG23498** [comment]
Hey @LukeWood we did not generate the files again. Went in to the md and ipynb and made the necessary updates. Hope this is alright 😄

**Comment by LukeWood** [comment]
works for me - thanks!!!

**Review by LukeWood — APPROVED** [review]

**Comment by LukeWood** [comment]
@fchollet for a last review

**Review by fchollet — COMMENTED** [review]
Thanks for the PR -- this is a very cool example! 👍  Thanks @LukeWood for the review too!

I pushed some copyedits, please pull them an regenerate the files. Then we can merge.

**Comment by ariG23498** [comment]
@LukeWood @fchollet I have pulled the copy edits and generated the files again. Thanks for the review 😄

**Comment by fchollet** [comment]
Thanks for the great contribution!
